cpp/src/io/fst/logical_stack.cuh

```diff
@@ -0,0 +1,523 @@
+  d_kv_operations_unsigned = cub::DoubleBuffer<StackOpUnsignedT>{
+    reinterpret_cast<StackOpUnsignedT*>(d_kv_operations.Current()),
+    reinterpret_cast<StackOpUnsignedT*>(d_kv_operations.Alternate())};
+  CUDF_CUDA_TRY(cub::DeviceRadixSort::SortPairs(temp_storage.data(),
+                                                total_temp_storage_bytes,
+                                                d_kv_operations_unsigned,
+                                                d_symbol_positions_db,
+                                                num_symbols_in,
+                                                begin_bit,
+                                                end_bit,
+                                                stream));
+
+  // TransformInputIterator that remaps all operations on stack level 0 to the empty stack symbol
+  kv_ops_scan_in  = {reinterpret_cast<StackOpT*>(d_kv_operations_unsigned.Current()),
+                    detail::RemapEmptyStack<StackOpT>{empty_stack}};
+  kv_ops_scan_out = reinterpret_cast<StackOpT*>(d_kv_operations_unsigned.Alternate());
+
+  // Dump info on stack operations sorted by their stack level (i.e. stack level after applying
+  // operation)
+  test::print::print_array(num_symbols_in,
+                           stream,
+                           get_stack_level_iterator(kv_ops_scan_in),
+                           get_value_iterator(kv_ops_scan_in));
+
+  // Inclusive scan to match pop operations with the latest push operation of that level
+  CUDF_CUDA_TRY(cub::DeviceScan::InclusiveScan(
+    temp_storage.data(),
+    total_temp_storage_bytes,
+    kv_ops_scan_in,
+    kv_ops_scan_out,
+    detail::PopulatePopWithPush<StackSymbolToStackOpTypeT>{symbol_to_stack_op},
+    num_symbols_in,
+    stream));
+
+  // Dump info on stack operations sorted by their stack level (i.e. stack level after applying
+  // operation)
+  test::print::print_array(num_symbols_in,
+                           stream,
+                           get_stack_level_iterator(kv_ops_scan_in),
+                           get_value_iterator(kv_ops_scan_in),
+                           get_stack_level_iterator(kv_ops_scan_out),
+                           get_value_iterator(kv_ops_scan_out));
+
+  // Fill the output tape with read-symbol
+  thrust::fill(thrust::cuda::par.on(stream),
+               thrust::device_ptr<StackSymbolT>{d_top_of_stack},
+               thrust::device_ptr<StackSymbolT>{d_top_of_stack + num_symbols_out},
+               read_symbol);
+
+  // Transform the stack operations to the stack symbol they represent
+  cub::TransformInputIterator<StackSymbolT, detail::StackOpToStackSymbol, StackOpT*>
+    kv_op_to_stack_sym_it(kv_ops_scan_out, detail::StackOpToStackSymbol{});
+
+  // Scatter the stack symbols to the output tape (spots that are not scattered to have been
+  // pre-filled with the read-symbol)
+  thrust::scatter(thrust::cuda::par.on(stream),
```

Comment:
```suggestion
  thrust::scatter(rmm::exec_policy(stream),
```
I suspect we can see `cudaMalloc` and `cudaFree` around this kernel in profiling since scatter definitely needs temporary memory storage

Reply:
Done 👍